### PR TITLE
Added some comments on usher_server

### PR DESCRIPTION
**Pull request by ShohM2**
**How to use usher_server**

1. Compilation
    -When compiling, add the CMake flag  -DUSHER_SERVER=TRUE

2. Input arguments for usher_server
    -a: argument directory. This directory would contain files in which arguments are contained.
    -i: list of mutation annotated trees(MATs). This file would contain the names of the protobuf files that 
              store MATs. Each line would have a name of one protobuf file.
    -s: sleep length in milliseconds. When there are no files in the argument directory, this is how long the 
         program will wait for before checking again. Default is 100 milliseconds
    -c: termination character. This is a character that will determine when the file in the argument directory 
        is ready to be read. This will be specified with the integer ASCII code value. When the specified
        character is at the end of the file, the arguments in the file will be read. Default is ‘^’ character.

3. Passing in usher arguments
- In the argument directory, add argument files that will contain arguments that are used to run usher with.
- While these files are being edited, the file will not be read. To run these arguments, add the termination character at the end.
- After the files are read, they are automatically deleted.

<ins>**Example**</ins>

`./usher_server -a example_arg_dir -s 50 -c 43 -i example_MAT_list.txt`

- This will run usher_server with an argument directory named example_arg_dir and a MAT list file named example_MAT_list.txt.
- Sleep time is set to 50 milliseconds and the termination character is ‘+’

- Contents of example_arg_dir
              example_arg1.txt,  example_arg2.txt, example_arg3.txt

<ins>example_MAT_list.txt</ins>
```
example_MAT1.pb
example_MAT2.pb
example_MAT3.pb
example_MAT4.pb
```

<ins>example_arg1.txt</ins>
```
-i example_MAT1.pb -v example/new_samples1.vcf -u -d output/+
```

<ins>example_arg2.txt</ins>
```
-i example_MAT2.pb -v example/new_samples2.vcf -u -k 20 -d output/
-i example_MAT3.pb -v example/new_samples3.vcf -u -d output/+
```

<ins>example_arg3.txt</ins>
```
-i example_MAT4.pb -v example/new_samples4.vcf -u -d output/
```

- In this example, arguments inside example_arg1.txt and example_arg2.txt will run since those files have the termination character while the argument inside example_arg3.txt will not run since example_arg3.txt does not contain the termination character.

- After that, example_arg1.txt and example_arg2.txt will be deleted. And, the program will wait until the termination character is added to the end of example_arg3.txt or a new file with the termination character is added to the directory.

### CODE REVIEW

File: src/usher_server.cpp [hunk 1]
```diff
@@ -64,13 +59,13 @@ int main(int argc, char** argv) {
...
 
-
+    //slot for a MAT not in the MAT_list file 
     MAT::Tree loaded_MAT;
     bool loaded_MAT_avail = false;//keep track if loaded_MAT is a new copy of MAT and can be used
     std::string loaded_MAT_name = "";
     
     fs::path p = fs::current_path();
-    p/=arg_dirname;
+    p/=arg_dirname;//get the path to the argument directory
```
Comment (yatisht) [inline src/usher_server.cpp:68]: I prefer not to have comments and code on the same lines. You should shift those comments one line above.

**Comment by yatisht** [comment]
Thanks. @AngieHinrichs, does this look good now?

**Comment by AngieHinrichs** [comment]
I wish I'd thought of this earlier, but there are several important things that I parse out of usher stderr output (Sample name, parsimony score, number of placements, imputed mutations) -- it would be a lot more convenient for those to be written to a tab-separated output file in the output directory, along with the other files, so I don't have to go through the whole stderr output of a long-running server process to find the appropriate lines of output for my latest request.  That's just an output formatting issue and I know exactly what I need for that, so I'm happy to submit a PR if that's OK with you @yatisht and @ShohM2.

Also, there are several error conditions that cause exit(), which is a good thing for a command-line program, but maybe not so good for a server process that is supposed to remain up and running.  It would be better for the server process to have a way of communicating the error condition back to the client without terminating.  That's a more interesting server architecture issue so I'm happy to leave that up to you guys, although I can make a suggestion if you'd like.

I like the configurable end-of-input marker in the arg files.  Is there a corresponding end-of-output signal in any of the output files?  I would like to be sure that when I read the output files, they are all complete.

**Comment by AngieHinrichs** [comment]
P.S. It doesn't have to be a character at the end of each file (in fact, parsing is easier without that); it could be the existence of a specially named file that indicates that all output files have been written and closed.

**Comment by russcd** [comment]
This is a really good idea, Angie. I wish we did like a year ago! 

It would be extra nice if that table could be formatted in an auspice.us compatible way so users can drag and drop the uncertainty output after placement. I'll add this as metadata in the snakemake workflow, too.

**Comment by AngieHinrichs** [comment]
Uncertainty output?  Is that number of placements?  I think auspice.us can take pretty much any TSV with the sample name in the first column.  (In fact, I don't know if this is on their main site yet, but they've recently added support for .XLS files too.)

**Comment by yatisht** [comment]
Hi @AngieHinrichs, Yes, I'm happy to merge your PR for the output file. 

I'll think of a way to throw exceptions back to the client instead of terminating with an exit().

**Comment by AngieHinrichs** [comment]
@ShohM2 in usher_server.cpp there's a vector of booleans, loaded_MAT_avail, corresponding to the vector of MAT::Tree, loaded_MAT.  It looks like when a loaded MAT is used for placement, the corresponding bit in loaded_MAT_avail is set to false.  Is that done in order to avoid using a loaded MAT more than once?  And if so, are we avoiding using a loaded MAT more than once because the MAT will have had samples from a previous request added to it, and/or some other reasons?  

As far as I can tell, a pre-loaded MAT is used only once.  The next time the same MAT is requested, it is reloaded from the protobuf file, which means that we're usually not getting the preloading speed advantage that motivated having a server process.  Is that correct or am I misunderstanding something?  

Perhaps instead of invalidating a MAT after it has been used once, we could make a copy of the MAT, and pass the copy to usher_common, leaving the preloaded MAT in its initial state, and then delete the copied MAT when done with it?

**Comment by ShohM2** [comment]
@AngieHinrichs Yes, this is done to avoid using a loaded MAT more than once since the MAT may be modified once they are used. The MATs that were used would be loaded again before going into waiting for more inputs. So when there is enough time to load, these MATs would be loaded in by the time more inputs come in. 

I initially tried copying the preloaded MATs, but the copying was taking longer than loading the MATs, so I decided to load every time.

**Comment by russcd** [comment]
There's no way to just prune added samples out of the tree? That function
is already in matutils and will be very fast to reset the tree.

**Pull request merged**